lib/src/iterable_extensions.dart

```diff
@@ -0,0 +1,725 @@
+  /// previous call, or the first element for the first call, and
+  /// the current element.
+  ///
+  /// Returns the result of the last call, or the first element if
+  /// there is only one element.
+  /// There must be at least one element.
+  T reduceIndexed(T Function(int index, T previous, T element) combine) {
+    var iterator = this.iterator;
+    if (!iterator.moveNext()) {
+      throw StateError('no elements');
+    }
+    var index = 1;
+    var result = iterator.current;
+    while (iterator.moveNext()) {
+      result = combine(index++, result, iterator.current);
+    }
+    return result;
+  }
+
+  /// Combine the elements with a value and the current index.
+  ///
+  /// Calls [combine] for each element with the current index,
+  /// the result of the previous call, or [initialValue] for the first element,
+  /// and the current element.
+  ///
+  /// Returns the result of the last call to [combine],
+  /// or [initialValue] if there are no elements.
+  R foldIndexed<R>(
+      R initialValue, R Function(int index, R previous, T element) combine) {
+    var result = initialValue;
+    var index = 0;
+    for (var element in this) {
+      result = combine(index++, result, element);
+    }
+    return result;
+  }
+
+  /// The first element satisfying [test], or `null` if there are none.
+  T? firstWhereOrNull(bool Function(T element) test) {
+    for (var element in this) {
+      if (test(element)) return element;
+    }
+    return null;
+  }
+
+  /// The first element whose value and index satisfies [test].
+  ///
+  /// Returns `null` if there are no element and index satisfying [test].
+  T? firstWhereIndexedOrNull(bool Function(int index, T element) test) {
+    var index = 0;
+    for (var element in this) {
+      if (test(index++, element)) return element;
+    }
+    return null;
+  }
+
+  /// The first element, or `null` if the iterable is empty.
+  T? get firstOrNull {
+    var iterator = this.iterator;
+    if (iterator.moveNext()) return iterator.current;
+    return null;
+  }
+
+  /// The last element satisfying [test], or `null` if there are none.
+  T? lastWhereOrNull(bool Function(T element) test) {
+    T? result;
+    for (var element in this) {
+      if (test(element)) result = element;
+    }
+    return result;
+  }
+
+  /// The last element whose index and value satisfies [test].
+  ///
+  /// Returns `null` if no element and index satisfies [test].
+  T? lastWhereIndexedOrNull(bool Function(int index, T element) test) {
+    T? result;
+    var index = 0;
+    for (var element in this) {
+      if (test(index++, element)) result = element;
+    }
+    return result;
+  }
+
+  /// The last element, or `null` if the iterable is empty.
+  T? get lastOrNull {
+    if (isEmpty) return null;
+    return last;
+  }
+
+  /// The single element satisfying [test].
+  ///
+  /// Returns `null` if there are either no elements
+  /// or more than one element satisfying [test].
```

Comment:
In the SDK I think the `singleWhere` methods throw if there is more than one match, and use `orElse` only for the no-match case. This is a different behavior that may be a better design, but also could be confusing.

If we do want this behavior, can we call out the difference specifically?